Dedicate one set of chat handlers per room

dlqqq opened the pull request:
## Description

- Makes chat handlers non-singletons in Jupyter AI. Each chat handler is dedicated to at most 1 `YChat` instance.
- Makes `YChat` an instance attribute instead of a local variable that gets forwarded across methods.
- Reverts most changes to chat handlers (except for the base class).

## Demo

https://github.com/user-attachments/assets/ecd8e027-b5da-4f49-9edf-6f0c5eb71d73

dlqqq committed:
create new set of chat handlers per room

dlqqq committed:
make YChat an instance attribute on BaseChatHandler

dlqqq committed:
revert changes to chat handlers

dlqqq committed:
pre-commit

brichet commented on packages/jupyter-ai/jupyter_ai/extension.py:
```diff
@@ -238,41 +232,72 @@ class AiExtension(ExtensionApp):
-            if chat is None:
-                return
+        # get YChat document associated with the room
+        ychat = await self.get_chat(data["room"])
```
```suggestion
        ychat = await self.get_chat(room_id)
```

brichet commented on packages/jupyter-ai/jupyter_ai/extension.py:
```diff
@@ -238,41 +232,72 @@ class AiExtension(ExtensionApp):
+        is cached, i.e. successive calls with the same room ID quickly return a
+        cached value.
+
+        TODO: Determine if get_chat() should ever fail under normal usage
+        scenarios. If not, we should just raise an exception if chat is `None`,
+        and indicate the return type as just `YChat` instead of
+        `Optional[YChat]`. This will simplify the code by removing redundant
+        null checks.
+        """
```
I agree with that, we should probably raise an error, it shouldn't happen.
Let's keep this note for a following PR.

dlqqq replied:
I've added this to my notes 👍

brichet commented on packages/jupyter-ai/jupyter_ai/extension.py:
```diff
@@ -238,41 +232,72 @@ class AiExtension(ExtensionApp):
+        handlers dedicated to the room identified by `<room_id>`.
+        """
+
+        self.ychats_by_room: Dict[str, YChat] = {}
+        """Cache of YChat instances, indexed by room ID."""
```
Do we need this cached dict?
If I'm not mistaken, this cache is accessed only in the `get_chat()`, which is called only when the room is initialized.

**Note**: For a following PR, if the `get_chat()` function is called only in `connect_chat()`, we should probably merge these functions.

dlqqq replied:
Originally, my code did use this dict, but it doesn't use this dict anymore. It can be removed later.

brichet commented on packages/jupyter-ai/jupyter_ai/extension.py:
```diff
@@ -238,41 +232,72 @@ class AiExtension(ExtensionApp):
+        `Optional[YChat]`. This will simplify the code by removing redundant
+        null checks.
+        """
+        if room_id in self.ychats_by_room:
```
See the comment above about `ychats_by_room`, but AFAIU this assessment should be always False.

brichet reviewed (commented):
Thanks @dlqq, this looks great and should considerably reduce the diffs in https://github.com/jupyterlab/jupyter-ai/pull/1043.

I have some comments/questions below.

dlqqq committed:
use room_id local var

Co-authored-by: Nicolas Brichet <32258950+brichet@users.noreply.github.com>

dlqqq commented:
@brichet Addressed your comments. As for the `ychats_by_room` dict, we can remove that later. It has a tiny memory footprint, and there could be a future use-case for caching `get_room()` calls.
